Components: Add a Focusable component for a button-like wrapper element

ryelle opened the pull request:
A few times (see #21753, #15940), it has come up that we want some reusable way to handle key events on components that we already have onClick events on. In the past, we've added `onClick` events to `div`s, which is not accessible by keyboard navigation, only to have to fix this later. This PR adds an accessible, button-like component that we can drop in place of those container divs.

This calls the onClick callback when the element is clicked, but it also adds a button role for screen readers, and a `tabIndex` so that it is focusable. When focused, this will also call the onClick callback when enter or space are pressed. This component adds almost no styles, only a focus style using the `accessible-focus` class. A dev can still pass an onKeyDown callback to catch events for all other key presses (arrow navigation, for example).

~This also adds a section on use to devdocs/button, and details to [the Button readme](https://github.com/Automattic/wp-calypso/tree/add/component-aria-button/client/components/button#blockbutton).~ 

This has been moved to its own component folder, so the devdocs example can be found at devdocs/focusable, and [the readme in components/focusable](https://github.com/Automattic/wp-calypso/tree/add/component-aria-button/client/components/focusable)

Devdocs demo with focus:

<img width="718" alt="screen shot 2018-02-16 at 11 10 45 am" src="https://user-images.githubusercontent.com/541093/36317074-12f02a08-130a-11e8-8d1e-5da134c2b7db.png">

### Questions ❓

- Should developers be able to pass through a different HTML element instead of divs? I think `div` catches almost all instances of this pattern.
- **Does this encourage a bad pattern (of not using `Button`)?** IMO, it's worse for an interactive element to be inaccessible, than a little janky - but would this experience be worse for anyone else?

ramonjd commented:
I think the examples on [/me/account](https://wordpress.com/me/account) (such as the _language_ and _main site_ settings components) warrant this accessibility sugar. As an enhancement to #21753, it will prevent some repetition.

> Does this encourage a bad pattern (of not using Button)? 

We'd have to limit ourselves to valid [child elements](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#Phrasing_content). Not impossible, but less flexible, especially since we could be importing further child components as children, which may contain non phrasing content. It'd make for a fair bit of refactoring.

gwwar commented on client/components/button/block.jsx:
```diff
@@ -0,0 +1,51 @@
+import { omit } from 'lodash';
+import PropTypes from 'prop-types';
+
+class BlockButton extends Component {
```
Given that this is a new component, I would expect this under `client/components/block-button/index.jsx` instead. (Searching is much easier if we're consistent with this.)

Do folks think otherwise?

gwwar replied:
Would a better name for this be something like `FocusableItem` or `Focusable` since it does look like we can take arbitrary children

ryelle replied:
I had the same thought when naming it, I was thinking block like "could contain block elements" which is definitely not clear. I like `Focusable` 👍

gwwar commented on client/components/button/README.md:
```diff
@@ -26,3 +26,29 @@ The following props are used to control the display of the component. Any additi
+
+## BlockButton
+
+This is a block-level button wrapper, using ARIA + keyboard handlers to create an element that can wrap other components. Buttons should only [contain phrasing content](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button), but there are places we have `div` wrappers with `onClick`. This is not accessible from the keyboard. This component should replace those divs, and adds button-lik ...
```
Could we link to what ARIA is? I don't think everyone would know.

gwwar commented on client/components/button/docs/example.jsx:
```diff
@@ -36,6 +38,22 @@ class Buttons extends React.PureComponent {
+					<div className="docs__design-button-row">
+						<BlockButton onClick={ noop }>
+							<p>
+								This "button" can contain other elements as children, including <code>p</code>s,{' '}
```
I'd recommend adding a _why_ explanation of when we'd use this.

gwwar replied:
Another section that has something like the items in https://github.com/Automattic/wp-calypso/pull/21753 would also give a better idea too

gwwar commented:
This looks promising 👍  I'd recommend moving this out of the button component folder though.

ryelle commented:
I've moved the component to `Focusable`, and updated the readme with hopefully clearer details on what it does and why that's useful for accessibility. I also added more specific instructions for better accessible alternatives, and why this pattern would be discouraged. 

> We'd have to limit ourselves to valid child elements. Not impossible, but less flexible.

Right, there can be UX tradeoffs for just using a button, but there are accessibility tradeoffs for using this method, too. I mentioned in the readme, but the entire text of the component is read out as the button name. Additionally, I'm not sure how a user using voice control would tell the computer to click this. This is what I meant by not wanting to encourage what's generally considered "bad practice".

gwwar commented on client/components/focusable/docs/example.jsx:
```diff
@@ -0,0 +1,66 @@
+			<Card>
+				<Focusable onClick={ noop }>
+					<p>
+						This keyboard-accessible component can contain other elements as children, including{' '}
```
I'd add a warning here too, and link back to the readme 👍

gwwar replied:
Oh I forgot that we show the readme now on http://calypso.localhost:3000/devdocs/design/focusable ! Could we maybe make the warning in the readme have some extra styling to make it stand out?

folletto commented:
Design wise seems a great idea, also seems matching our patterns, so let's go! 👍
